File: 15-3sum/15-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        
        for i,v in enumerate(nums):
            if i>0 and v == nums[i-1]:
                continue
            l,r = i+1, len(nums)-1
            while l<r:
                s = v + nums[l] + nums[r]
                if s > 0:
                    r-=1
                elif s <0:
                    l+=1
                else:
                    res.append([v,nums[l],nums[r]])
                    l+=1
                    while nums[l] == nums[l-1] and l<r:
                        l+=1
        return res

File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        
        Arrays.sort(nums);
        for(int i = 0; i<nums.length;i++){
            if (i >0 && nums[i] == nums[i-1]){
                continue;
            }
            int l = i+1;
            int r = nums.length -1;
            while (l<r){
                int sum = nums[l] + nums[r] + nums[i];
                if (sum <0){
                    l++;
                }
                else if (sum>0){
                    r--;
                }
                else {
                    int[] toAdd = {nums[i],nums[l],nums[r]};
                    List<Integer> list = Arrays.stream(toAdd).boxed().collect(Collectors.toList());
                    res.add(list);
                    l++;
                    while (nums[l] == nums[l-1] && l<r){
                        l++;
                    }
                }
                
            }
        }
        return res;
    }
}